index.js

```diff
@@ -16,65 +20,67 @@ util.inherits(CpFileError, NestedError);
 
 CpFileError.prototype.name = 'CpFileError';
 
-module.exports = function (src, dest, opts, cb) {
+module.exports = function (src, dest, opts) {
 	if (!src || !dest) {
 		throw new CpFileError('`src` and `dest` required');
 	}
 
 	if (typeof opts !== 'object') {
-		cb = opts;
 		opts = {};
 	}
 
-	cb = onetime(cb || function () {});
 	opts = objectAssign({overwrite: true}, opts);
 
-	var read = fs.createReadStream(src);
-	var readListener = onetime(startWrite);
+	return new Promise(function startRead(resolve, reject) {
+		var read = fs.createReadStream(src);
 
-	read.on('error', function (err) {
-		cb(new CpFileError('cannot read from `' + src + '`: ' + err.message, err));
-	});
-	read.on('readable', readListener);
-	read.on('end', readListener);
-
-	function startWrite() {
-		mkdirp(path.dirname(dest), function (err) {
-			if (err && err.code !== 'EEXIST') {
-				cb(new CpFileError('cannot create directory `' + path.dirname(dest) + '`: ' + err.message, err));
-				return;
+		read.on('error', function (err) {
+			reject(new CpFileError('cannot read from `' + src + '`: ' + err.message, err));
+		});
+		read.on('readable', function () {
+			resolve(read);
+		});
+		read.on('end', function () {
+			resolve(read);
+		});
+	}).then(function mkdirpDestDirectory(read) {
+		return mkdirpP(path.dirname(dest)).then(function () {
+			return read;
+		}).catch(function (err) {
+			if (err.code !== 'EEXIST') {
+				throw new CpFileError('cannot create directory `' + path.dirname(dest) + '`: ' + err.message, err);
 			}
-
+			return read;
+		});
+	}).then(function (read) {
+		return new Promise(function pipeToDest(resolve, reject) {
 			var write = fs.createWriteStream(dest, {flags: opts.overwrite ? 'w' : 'wx'});
 
 			write.on('error', function (err) {
 				if (!opts.overwrite && err.code === 'EEXIST') {
-					cb();
+					resolve(false);
 					return;
 				}
-				cb(new CpFileError('cannot write to `' + dest + '`: ' + err.message, err));
+				reject(new CpFileError('cannot write to `' + dest + '`: ' + err.message, err));
 			});
 
 			write.on('close', function () {
-				fs.lstat(src, function (err, stats) {
-					if (err) {
-						cb(new CpFileError('lstat `' + src + '` failed: ' + err.message, err));
-						return;
-					}
-
-					fs.utimes(dest, stats.atime, stats.mtime, function (err) {
-						if (err) {
-							cb(new CpFileError('utimes `' + dest + '` failed: ' + err.message, err));
-							return;
-						}
-						cb();
-					});
-				});
+				resolve(true);
 			});
 
 			read.pipe(write);
 		});
-	}
+	}).then(function (updateTimes) {
+		if (updateTimes) {
+			return lstatP(src).catch(function (err) {
+				throw new CpFileError('lstat `' + src + '` failed: ' + err.message, err);
+			}).then(function (stats) {
+				return utimesP(dest, stats.atime, stats.mtime).catch(function (err) {
+					throw new CpFileError('utimes `' + dest + '` failed: ' + err.message, err);
+				});
+			});
```

Comment:
@SamVerschueren Sorry, I don't know if i understood you correctly. Would you prefer that flow, or do you think that is the flow of my PR?